src/main/java/org/apache/sysml/runtime/codegen/LibSpoofPrimitives.java

```diff
@@ -1379,7 +1379,72 @@ public static void vectMult2Add(double[] a, double[] c, int[] aix, int ai, int c
 			c[aix[j]] = a[j] + a[j];
 		return c;
 	}
-	
+
+	//custom vector xor
+    //TODO sparse operations are not working correctly
+	public static void vectXorAdd(double[] a, double bval, double[] c, int ai, int ci, int len) {
+		for( int j = ai; j < ai+len; j++, ci++)
+			c[ci] +=  ( (a[j] != 0) != (bval != 0) ) ? 1 : 0;
+	}
+
+	public static void vectXorAdd(double bval, double[] a, double[] c, int ai, int ci, int len) {
+		for( int j = ai; j < ai+len; j++, ci++)
+			c[ci] += ( (bval != 0) != (a[j] != 0) ) ? 1 : 0;
+	}
+
+	public static void vectXorAdd(double[] a, double bval, double[] c, int[] aix, int ai, int ci, int alen, int len) {
+		for( int j = ai; j < ai+alen; j++ )
+			c[ci + aix[j]] += ( a[j] != bval ) ? 1 : 0;
+	}
+
+	public static void vectXorAdd(double bval, double[] a, double[] c, int[] aix, int ai, int ci, int alen, int len) {
+		for( int j = ai; j < ai+alen; j++ )
+			c[ci + aix[j]] += ( bval != a[j] ) ? 1 : 0;
+	}
+
+	public static double[] vectXorWrite(double[] a, double bval, int ai, int len) {
+		double[] c = allocVector(len, false);
+		for( int j = 0; j < len; j++)
+			c[j] = ( ( a[ai+j] != 0) != (bval != 0) ) ? 1 : 0;
+		return c;
+	}
+
+	public static double[] vectXorWrite(double bval, double[] a, int ai, int len) {
+		double[] c = allocVector(len, false);
+		for( int j = 0; j < len; j++)
+			c[j] = ( (bval != 0) != (a[ai + j] != 0) ) ? 1 : 0;
+		return c;
+	}
+
+	public static double[] vectXorWrite(double[] a, double[] b, int ai, int bi, int len) {
+		double[] c = allocVector(len, false);
+		for( int j = 0; j < len; j++)
+			c[j] = ( (a[ai + j] != 0) != (b[bi + j] != 0) ) ? 1 : 0;
+		return c;
+	}
+
+	public static double[] vectXorWrite(double[] a, double bval, int[] aix, int ai, int alen, int len) {
+		//double init = (bval != 0) ? 0 : Double.NaN;
+		double[] c = allocVector(len, false);
+		for( int j = ai; j < ai+alen; j++ )
+			c[aix[j]] = ( (a[j] != 0) != (bval != 0) ) ? 1 : 0;
+		return c;
+	}
+
+	public static double[] vectXorWrite(double bval, double[] a, int[] aix, int ai, int alen, int len) {
+		double[] c = allocVector(len, true, 1);
```

Comment:
This initialization is incorrect - we want to initialize to xor(bval, 0) to capture all values that are not existing in the sparse row a.